File: ThreeDSecure/src/test/java/com/braintreepayments/api/threedsecure/ThreeDSecureNonceUnitTest.java
package com.braintreepayments.api.threedsecure;

import static com.braintreepayments.api.testutils.Assertions.assertBinDataEqual;
import static com.braintreepayments.api.card.BinData.NO;
import static com.braintreepayments.api.card.BinData.UNKNOWN;
import static com.braintreepayments.api.card.BinData.YES;
import static junit.framework.Assert.assertEquals;
import static junit.framework.Assert.assertFalse;
import static junit.framework.Assert.assertNotNull;
import static junit.framework.Assert.assertTrue;
import static org.junit.Assert.assertNull;

import android.os.Parcel;

import com.braintreepayments.api.testutils.Fixtures;

import org.json.JSONException;
import org.json.JSONObject;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;

@RunWith(RobolectricTestRunner.class)
public class ThreeDSecureNonceUnitTest {
    @Test
    public void fromJSON_parsesThreeDSecureNonce() throws JSONException {
        ThreeDSecureNonce threeDSecureNonce = ThreeDSecureNonce.fromJSON(new JSONObject(Fixtures.PAYMENT_METHODS_RESPONSE_VISA_CREDIT_CARD));

        assertEquals("Visa", threeDSecureNonce.getCardType());
        assertEquals("123456-12345-12345-a-adfa", threeDSecureNonce.getString());
        assertEquals("11", threeDSecureNonce.getLastTwo());
        assertEquals("1111", threeDSecureNonce.getLastFour());
        assertNotNull(threeDSecureNonce.getThreeDSecureInfo());
        assertFalse(threeDSecureNonce.getThreeDSecureInfo().getLiabilityShifted());
        assertTrue(threeDSecureNonce.getThreeDSecureInfo().getLiabilityShiftPossible());
        assertNotNull(threeDSecureNonce.getBinData());
        assertEquals(UNKNOWN, threeDSecureNonce.getBinData().getPrepaid());
        assertEquals(YES, threeDSecureNonce.getBinData().getHealthcare());
        assertEquals(NO, threeDSecureNonce.getBinData().getDebit());
        assertEquals(UNKNOWN, threeDSecureNonce.getBinData().getDurbinRegulated());
        assertEquals(UNKNOWN, threeDSecureNonce.getBinData().getCommercial());
        assertEquals(UNKNOWN, threeDSecureNonce.getBinData().getPayroll());
        assertEquals(UNKNOWN, threeDSecureNonce.getBinData().getIssuingBank());
        assertEquals("Something", threeDSecureNonce.getBinData().getCountryOfIssuance());
        assertEquals("123", threeDSecureNonce.getBinData().getProductId());
        assertEquals("unregulated",
                threeDSecureNonce.getAuthenticationInsight().getRegulationEnvironment());
        assertEquals("01", threeDSecureNonce.getExpirationMonth());
        assertEquals("2020", threeDSecureNonce.getExpirationYear());
        assertEquals("Joe Smith", threeDSecureNonce.getCardholderName());
    }

    @Test
    public void fromJSON_withGraphQLTokenizationResponse_parsesCardNonce() throws JSONException {
        ThreeDSecureNonce threeDSecureNonce = ThreeDSecureNonce.fromJSON(new JSONObject(Fixtures.GRAPHQL_RESPONSE_CREDIT_CARD));

        assertEquals("Visa", threeDSecureNonce.getCardType());
        assertEquals("3744a73e-b1ab-0dbd-85f0-c12a0a4bd3d1", threeDSecureNonce.getString());
        assertEquals("11", threeDSecureNonce.getLastTwo());
        assertEquals("1111", threeDSecureNonce.getLastFour());
        assertNotNull(threeDSecureNonce.getThreeDSecureInfo());
        assertFalse(threeDSecureNonce.getThreeDSecureInfo().getLiabilityShifted());
        assertFalse(threeDSecureNonce.getThreeDSecureInfo().getLiabilityShiftPossible());
        assertNotNull(threeDSecureNonce.getBinData());
        assertEquals(YES, threeDSecureNonce.getBinData().getPrepaid());
        assertEquals(YES, threeDSecureNonce.getBinData().getHealthcare());
        assertEquals(NO, threeDSecureNonce.getBinData().getDebit());
        assertEquals(YES, threeDSecureNonce.getBinData().getDurbinRegulated());
        assertEquals(NO, threeDSecureNonce.getBinData().getCommercial());
        assertEquals(YES, threeDSecureNonce.getBinData().getPayroll());
        assertEquals("Bank of America", threeDSecureNonce.getBinData().getIssuingBank());
        assertEquals("USA", threeDSecureNonce.getBinData().getCountryOfIssuance());
        assertEquals("123", threeDSecureNonce.getBinData().getProductId());
        assertEquals("unregulated",
                threeDSecureNonce.getAuthenticationInsight().getRegulationEnvironment());
        assertEquals("01", threeDSecureNonce.getExpirationMonth());
        assertEquals("2020", threeDSecureNonce.getExpirationYear());
        assertEquals("Joe Smith", threeDSecureNonce.getCardholderName());
    }

    @Test
    public void fromJSON_withGraphQLTokenizationResponse_parsesThreeDSecureWithDefaultValues() throws JSONException {
        ThreeDSecureNonce threeDSecureNonce = ThreeDSecureNonce.fromJSON(new JSONObject(Fixtures.GRAPHQL_RESPONSE_CREDIT_CARD_MISSING_VALUES));

        assertEquals("", threeDSecureNonce.getLastFour());
        assertEquals("", threeDSecureNonce.getLastTwo());
        assertEquals("Unknown", threeDSecureNonce.getCardType());
        assertNotNull(threeDSecureNonce.getThreeDSecureInfo());
        assertEquals("", threeDSecureNonce.getBin());
        assertNotNull(threeDSecureNonce.getBinData());
        assertEquals("3744a73e-b1ab-0dbd-85f0-c12a0a4bd3d1", threeDSecureNonce.getString());
        assertFalse(threeDSecureNonce.isDefault());
        assertNull(threeDSecureNonce.getAuthenticationInsight());
        assertEquals("", threeDSecureNonce.getExpirationMonth());
        assertEquals("", threeDSecureNonce.getExpirationYear());
        assertEquals("", threeDSecureNonce.getCardholderName());
    }

    @Test
    public void fromJSON_withGraphQLTokenizationResponse_parsesUnknownCardResponses() throws JSONException {
        ThreeDSecureNonce threeDSecureNonce = ThreeDSecureNonce.fromJSON(new JSONObject(Fixtures.GRAPHQL_RESPONSE_UNKNOWN_CREDIT_CARD));

        assertEquals("Unknown", threeDSecureNonce.getCardType());
        assertEquals("tokencc_3bbd22_fpjshh_bqbvh5_mkf3nf_smz", threeDSecureNonce.getString());
        assertEquals("", threeDSecureNonce.getLastTwo());
        assertEquals("", threeDSecureNonce.getLastFour());
        assertEquals("", threeDSecureNonce.getExpirationMonth());
        assertEquals("", threeDSecureNonce.getExpirationYear());
        assertEquals("", threeDSecureNonce.getCardholderName());
        assertNotNull(threeDSecureNonce.getThreeDSecureInfo());
        assertFalse(threeDSecureNonce.getThreeDSecureInfo().getLiabilityShifted());
        assertFalse(threeDSecureNonce.getThreeDSecureInfo().getLiabilityShiftPossible());
        assertNotNull(threeDSecureNonce.getBinData());
        assertEquals(UNKNOWN, threeDSecureNonce.getBinData().getPrepaid());
        assertEquals(UNKNOWN, threeDSecureNonce.getBinData().getHealthcare());
        assertEquals(UNKNOWN, threeDSecureNonce.getBinData().getDebit());
        assertEquals(UNKNOWN, threeDSecureNonce.getBinData().getDurbinRegulated());
        assertEquals(UNKNOWN, threeDSecureNonce.getBinData().getCommercial());
        assertEquals(UNKNOWN, threeDSecureNonce.getBinData().getPayroll());
        assertEquals(UNKNOWN, threeDSecureNonce.getBinData().getIssuingBank());
        assertEquals(UNKNOWN, threeDSecureNonce.getBinData().getCountryOfIssuance());
        assertEquals(UNKNOWN, threeDSecureNonce.getBinData().getProductId());
    }

    @Test
    public void parcelsCorrectly() throws JSONException {
        ThreeDSecureNonce threeDSecureNonce = ThreeDSecureNonce.fromJSON(new JSONObject(Fixtures.PAYMENT_METHODS_RESPONSE_VISA_CREDIT_CARD));

        Parcel parcel = Parcel.obtain();
        threeDSecureNonce.writeToParcel(parcel, 0);
        parcel.setDataPosition(0);

        ThreeDSecureNonce parceled = ThreeDSecureNonce.CREATOR.createFromParcel(parcel);

        assertEquals("Visa", parceled.getCardType());
        assertEquals("123456-12345-12345-a-adfa", parceled.getString());
        assertEquals("11", parceled.getLastTwo());
        assertEquals("1111", parceled.getLastFour());
        assertEquals("01", parceled.getExpirationMonth());
        assertEquals("2020", parceled.getExpirationYear());
        assertEquals("Joe Smith", parceled.getCardholderName());
        assertFalse(parceled.isDefault());
        assertBinDataEqual(threeDSecureNonce.getBinData(), parceled.getBinData());
        assertEquals(threeDSecureNonce.getAuthenticationInsight().getRegulationEnvironment(),
                parceled.getAuthenticationInsight().getRegulationEnvironment());
        assertEquals(threeDSecureNonce.getThreeDSecureInfo().getLiabilityShifted(), parceled.getThreeDSecureInfo().getLiabilityShifted());
        assertEquals(threeDSecureNonce.getThreeDSecureInfo().getLiabilityShiftPossible(), parceled.getThreeDSecureInfo().getLiabilityShiftPossible());
    }
}

File: ThreeDSecure/src/test/java/com/braintreepayments/api/threedsecure/ThreeDSecureNonceUnitTest.kt
package com.braintreepayments.api.threedsecure

import android.os.Parcel
import com.braintreepayments.api.card.BinType
import com.braintreepayments.api.testutils.Assertions.assertBinDataEqual
import com.braintreepayments.api.testutils.Fixtures
import kotlinx.parcelize.parcelableCreator
import org.json.JSONObject
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertNotNull
import kotlin.test.assertNull
import kotlin.test.assertTrue

@RunWith(RobolectricTestRunner::class)
class ThreeDSecureNonceUnitTest {

    @Test
    fun `fromJSON parses ThreeDSecureNonce`() {
        val sut = ThreeDSecureNonce.fromJSON(JSONObject(Fixtures.PAYMENT_METHODS_RESPONSE_VISA_CREDIT_CARD))

        assertEquals("Visa", sut.cardType)
        assertEquals("123456-12345-12345-a-adfa", sut.string)
        assertEquals("11", sut.lastTwo)
        assertEquals("1111", sut.lastFour)
        assertNotNull(sut.threeDSecureInfo)
        assertFalse { sut.threeDSecureInfo.liabilityShifted }
        assertTrue { sut.threeDSecureInfo.liabilityShiftPossible }
        assertNotNull(sut.binData)
        assertEquals(BinType.Unknown, sut.binData.prepaid)
        assertEquals(BinType.Yes, sut.binData.healthcare)
        assertEquals(BinType.No, sut.binData.debit)
        assertEquals(BinType.Unknown, sut.binData.durbinRegulated)
        assertEquals(BinType.Unknown, sut.binData.commercial)
        assertEquals(BinType.Unknown, sut.binData.payroll)
        assertEquals(BinType.Unknown.name, sut.binData.issuingBank)
        assertEquals("Something", sut.binData.countryOfIssuance)
        assertEquals("123", sut.binData.productId)
        assertEquals("unregulated", sut.authenticationInsight?.regulationEnvironment)
        assertEquals("01", sut.expirationMonth)
        assertEquals("2020", sut.expirationYear)
        assertEquals("Joe Smith", sut.cardholderName)
    }

    @Test
    fun `fromJSON with GraphQL tokenization response parses CardNonce`() {
        val sut = ThreeDSecureNonce.fromJSON(JSONObject(Fixtures.GRAPHQL_RESPONSE_CREDIT_CARD))

        assertEquals("Visa", sut.cardType)
        assertEquals("3744a73e-b1ab-0dbd-85f0-c12a0a4bd3d1", sut.string)
        assertEquals("11", sut.lastTwo)
        assertEquals("1111", sut.lastFour)
        assertNotNull(sut.threeDSecureInfo)
        assertFalse { sut.threeDSecureInfo.liabilityShifted }
        assertFalse { sut.threeDSecureInfo.liabilityShiftPossible }
        assertNotNull(sut.binData)
        assertEquals(BinType.Yes, sut.binData.prepaid)
        assertEquals(BinType.Yes, sut.binData.healthcare)
        assertEquals(BinType.No, sut.binData.debit)
        assertEquals(BinType.Yes, sut.binData.durbinRegulated)
        assertEquals(BinType.No, sut.binData.commercial)
        assertEquals(BinType.Yes, sut.binData.payroll)
        assertEquals("Bank of America", sut.binData.issuingBank)
        assertEquals("USA", sut.binData.countryOfIssuance)
        assertEquals("123", sut.binData.productId)
        assertEquals("unregulated", sut.authenticationInsight?.regulationEnvironment)
        assertEquals("01", sut.expirationMonth)
        assertEquals("2020", sut.expirationYear)
        assertEquals("Joe Smith", sut.cardholderName)
    }

    @Test
    fun `fromJSON with GraphQL tokenization response parses ThreeDSecure with default values`() {
        val sut = ThreeDSecureNonce.fromJSON(JSONObject(Fixtures.GRAPHQL_RESPONSE_CREDIT_CARD_MISSING_VALUES))

        assertEquals("", sut.lastFour)
        assertEquals("", sut.lastTwo)
        assertEquals("Unknown", sut.cardType)
        assertNotNull(sut.threeDSecureInfo)
        assertEquals("", sut.bin)
        assertNotNull(sut.binData)
        assertEquals("3744a73e-b1ab-0dbd-85f0-c12a0a4bd3d1", sut.string)
        assertFalse(sut.isDefault)
        assertNull(sut.authenticationInsight)
        assertEquals("", sut.expirationMonth)
        assertEquals("", sut.expirationYear)
        assertEquals("", sut.cardholderName)
    }

    @Test
    fun `fromJSON with GraphQL tokenization response parses unknown card responses`() {
        val sut = ThreeDSecureNonce.fromJSON(JSONObject(Fixtures.GRAPHQL_RESPONSE_UNKNOWN_CREDIT_CARD))

        assertEquals("Unknown", sut.cardType)
        assertEquals("tokencc_3bbd22_fpjshh_bqbvh5_mkf3nf_smz", sut.string)
        assertEquals("", sut.lastTwo)
        assertEquals("", sut.lastFour)
        assertEquals("", sut.expirationMonth)
        assertEquals("", sut.expirationYear)
        assertEquals("", sut.cardholderName)
        assertNotNull(sut.threeDSecureInfo)
        assertFalse(sut.threeDSecureInfo.liabilityShifted)
        assertFalse(sut.threeDSecureInfo.liabilityShiftPossible)
        assertNotNull(sut.binData)
        assertEquals(BinType.Unknown, sut.binData.prepaid)
        assertEquals(BinType.Unknown, sut.binData.healthcare)
        assertEquals(BinType.Unknown, sut.binData.debit)
        assertEquals(BinType.Unknown, sut.binData.durbinRegulated)
        assertEquals(BinType.Unknown, sut.binData.commercial)
        assertEquals(BinType.Unknown, sut.binData.payroll)
        assertEquals(BinType.Unknown.name, sut.binData.issuingBank)
        assertEquals(BinType.Unknown.name, sut.binData.countryOfIssuance)
        assertEquals(BinType.Unknown.name, sut.binData.productId)
    }

    @Test
    fun `parcels correctly`() {
        val sut = ThreeDSecureNonce.fromJSON(JSONObject(Fixtures.PAYMENT_METHODS_RESPONSE_VISA_CREDIT_CARD))

        val parcel = Parcel.obtain().apply {
            sut.writeToParcel(this, 0)
            setDataPosition(0)
        }
        val parceled = parcelableCreator<ThreeDSecureNonce>().createFromParcel(parcel)

        assertEquals("Visa", parceled.cardType)
        assertEquals("123456-12345-12345-a-adfa", parceled.string)
        assertEquals("11", parceled.lastTwo)
        assertEquals("1111", parceled.lastFour)
        assertEquals("01", parceled.expirationMonth)
        assertEquals("2020", parceled.expirationYear)
        assertEquals("Joe Smith", parceled.cardholderName)
        assertFalse(parceled.isDefault)
        assertBinDataEqual(sut.binData, parceled.binData)
        assertEquals(sut.authenticationInsight?.regulationEnvironment,
            parceled.authenticationInsight?.regulationEnvironment)
        assertEquals(sut.threeDSecureInfo.liabilityShifted, parceled.threeDSecureInfo.liabilityShifted)
        assertEquals(sut.threeDSecureInfo.liabilityShiftPossible, parceled.threeDSecureInfo.liabilityShiftPossible)
    }
}
